Brokerages/Bitfinex/BitfinexBrokerage.Util.cs

```diff
@@ -0,0 +1,164 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+using QuantConnect.Orders;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using TradingApi.ModelObjects.Bitfinex.Json;
+
+namespace QuantConnect.Brokerages.Bitfinex
+{
+
+    public partial class BitfinexBrokerage
+    {
+
+        #region Declarations
+        private class OrderTypeMap
+        {
+            public string BitfinexOrderType { get; set; }
+            public string Wallet { get; set; }
+            public OrderType OrderType { get; set; }
+        }
+
+        enum WalletType
+        {
+            exchange,
+            trading
+        }
+
+        //todo: trailing stop support
+        private static List<OrderTypeMap> _orderTypeMap = new List<OrderTypeMap>
+        {
+            { new OrderTypeMap { BitfinexOrderType = "exchange market", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.Market } },
+            { new OrderTypeMap { BitfinexOrderType = "exchange limit", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.Limit } },
+            { new OrderTypeMap { BitfinexOrderType = "exchange stop", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.StopMarket } },
+            //{ new OrderTypeMap { BitfinexOrderType = "exchange trailing stop", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.StopLimit } },
+
+            { new OrderTypeMap { BitfinexOrderType = "market", Wallet = WalletType.trading.ToString(), OrderType = OrderType.Market } },
+            { new OrderTypeMap { BitfinexOrderType = "limit", Wallet = WalletType.trading.ToString(), OrderType = OrderType.Limit } },
+            { new OrderTypeMap { BitfinexOrderType = "stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopMarket } },
+            //{ new OrderTypeMap { BitfinexOrderType = "trailing stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopLimit } },
+        };
+        #endregion
+
+
+        /// <summary>
+        /// Map exchange status
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static OrderStatus MapOrderStatus(BitfinexOrderStatusResponse response)
+        {
+            decimal remainingAmount;
+            decimal executedAmount;
+            if (response.IsCancelled)
+            {
+                return OrderStatus.Canceled;
+            }
+            else if (decimal.TryParse(response.RemainingAmount, out remainingAmount) && remainingAmount > 0
+                && decimal.TryParse(response.ExecutedAmount, out executedAmount) && executedAmount > 0)
+            {
+                return OrderStatus.PartiallyFilled;
```

Comment:
Possibly. The Implemenatation will currently return Filled status when the order fee is returned, and this has proven to be reliable. I see no reason why an update with the fee could not be subsequent to zeroing the RemainingAmount where there are multiple trade executed messages, but I think it unlikely or impossible that the reverse may happen.